// saihubot-skill-search/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "skillSearchDuckduckgo", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchDuckduckgo;
  }
});
Object.defineProperty(exports, "skillSearchWikipedia", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchWikipedia;
  }
});
Object.defineProperty(exports, "skillSearchGoogle", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGoogle;
  }
});
Object.defineProperty(exports, "skillSearchGoogletranslate", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGoogletranslate;
  }
});
Object.defineProperty(exports, "skillSearchGoogleimages", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGoogleimages;
  }
});
Object.defineProperty(exports, "skillSearchGooglemaps", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGooglemaps;
  }
});
Object.defineProperty(exports, "skillSearchGoogleTrends", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGoogleTrends;
  }
});
Object.defineProperty(exports, "skillSearchYoutube", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchYoutube;
  }
});
Object.defineProperty(exports, "skillSearchTwitter", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchTwitter;
  }
});
Object.defineProperty(exports, "skillSearchWolfram", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchWolfram;
  }
});
Object.defineProperty(exports, "searchSkills", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skills;
  }
});
Object.defineProperty(exports, "skillSearchGithub", {
  enumerable: true,
  get: function () {
    return _saihubotSkillDevSearch.skillSearchGithub;
  }
});
Object.defineProperty(exports, "skillSearchMdn", {
  enumerable: true,
  get: function () {
    return _saihubotSkillDevSearch.skillSearchMdn;
  }
});
Object.defineProperty(exports, "skillSearchNpm", {
  enumerable: true,
  get: function () {
    return _saihubotSkillDevSearch.skillSearchNpm;
  }
});
Object.defineProperty(exports, "skillSearchStackoverflow", {
  enumerable: true,
  get: function () {
    return _saihubotSkillDevSearch.skillSearchStackoverflow;
  }
});
Object.defineProperty(exports, "skillSearchCheatsh", {
  enumerable: true,
  get: function () {
    return _saihubotSkillDevSearch.skillSearchCheatsh;
  }
});
Object.defineProperty(exports, "devSkills", {
  enumerable: true,
  get: function () {
    return _saihubotSkillDevSearch.skills;
  }
});
exports.skills = void 0;

var _saihubotSkillSearch = require("./saihubot-skill-search.js");

var _saihubotSkillDevSearch = require("./saihubot-skill-dev-search.js");

const skills = [..._saihubotSkillSearch.skills, ..._saihubotSkillDevSearch.skills];
exports.skills = skills;
