// saihubot-skill-search/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "skillSearchDuckduckgo", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchDuckduckgo;
  }
});
Object.defineProperty(exports, "skillSearchWikipedia", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchWikipedia;
  }
});
Object.defineProperty(exports, "skillSearchGithub", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGithub;
  }
});
Object.defineProperty(exports, "skillSearchGoogle", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGoogle;
  }
});
Object.defineProperty(exports, "skillSearchGoogletranslate", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGoogletranslate;
  }
});
Object.defineProperty(exports, "skillSearchGoogleimages", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGoogleimages;
  }
});
Object.defineProperty(exports, "skillSearchGooglemaps", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchGooglemaps;
  }
});
Object.defineProperty(exports, "skillSearchYoutube", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchYoutube;
  }
});
Object.defineProperty(exports, "skillSearchMdn", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchMdn;
  }
});
Object.defineProperty(exports, "skillSearchNpm", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchNpm;
  }
});
Object.defineProperty(exports, "skillSearchTwitter", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchTwitter;
  }
});
Object.defineProperty(exports, "skillSearchStackoverflow", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchStackoverflow;
  }
});
Object.defineProperty(exports, "skillSearchWolfram", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skillSearchWolfram;
  }
});
Object.defineProperty(exports, "skills", {
  enumerable: true,
  get: function () {
    return _saihubotSkillSearch.skills;
  }
});

var _saihubotSkillSearch = require("./saihubot-skill-search.js");
