// saihubot-cli-adapter/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "cliAdapter", {
  enumerable: true,
  get: function () {
    return _saihubotCliAdapter.default;
  }
});
Object.defineProperty(exports, "addonConfirm", {
  enumerable: true,
  get: function () {
    return _saihubotCliAddonDialog.addonConfirm;
  }
});
Object.defineProperty(exports, "addonSearch", {
  enumerable: true,
  get: function () {
    return _saihubotCliAddonSearch.addonSearch;
  }
});
Object.defineProperty(exports, "addonOpenLink", {
  enumerable: true,
  get: function () {
    return _saihubotCliAddonSearch.addonOpenLink;
  }
});
Object.defineProperty(exports, "addonFetch", {
  enumerable: true,
  get: function () {
    return _saihubotCliAddonSearch.addonFetch;
  }
});
Object.defineProperty(exports, "addonExec", {
  enumerable: true,
  get: function () {
    return _saihubotCliAddonSearch.addonExec;
  }
});
Object.defineProperty(exports, "skillHelp", {
  enumerable: true,
  get: function () {
    return _saihubotCliSkillHelp.skillHelp;
  }
});
Object.defineProperty(exports, "getConfig", {
  enumerable: true,
  get: function () {
    return _utils.getConfig;
  }
});
Object.defineProperty(exports, "t", {
  enumerable: true,
  get: function () {
    return _i18n.t;
  }
});
exports.skills = exports.addons = void 0;

var _saihubotCliAdapter = _interopRequireDefault(require("./saihubot-cli-adapter"));

var _saihubotCliAddonDialog = require("./saihubot-cli-addon-dialog.js");

var _saihubotCliAddonSearch = require("./saihubot-cli-addon-search");

var _saihubotCliSkillHelp = require("./saihubot-cli-skill-help");

var _utils = require("./utils");

var _i18n = require("./i18n");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const addons = [_saihubotCliAddonDialog.addonConfirm, _saihubotCliAddonSearch.addonSearch, _saihubotCliAddonSearch.addonOpenLink, _saihubotCliAddonSearch.addonFetch, _saihubotCliAddonSearch.addonExec];
exports.addons = addons;
const skills = [_saihubotCliSkillHelp.skillHelp];
exports.skills = skills;
