// staker/dist/saihubot-cli-skill-chainid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.skills = exports.skillChainId = void 0;

var _react = _interopRequireDefault(require("react"));

var _inkTable = _interopRequireDefault(require("ink-table"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Randomly picking an awesome ethereum tool.
 */
const skillChainId = {
  name: 'chainId',
  help: '📕chainid - Show EVM network providers chain and network id table',
  requirements: {
    addons: ['fetch']
  },
  rule: /^chainid$/i,
  action: function (robot, msg) {
    robot.send('Fetching chain ids...');
    robot.render();
    robot.addons.fetch('https://chainid.network/chains.json').then(response => response.json()).then(json => {
      const data = json.map(entry => ({
        name: entry.name,
        rpc: entry.rpc && (0, _utils.getRandomItem)(entry.rpc) || '',
        chainId: entry.chainId,
        symbol: entry.nativeCurrency && entry.nativeCurrency.symbol || ''
      })).filter(entry => entry.rpc !== '');
      robot.sendComponent( /*#__PURE__*/_react.default.createElement(_inkTable.default, {
        data: data
      }));
      robot.render();
    });
  }
};
exports.skillChainId = skillChainId;
const skills = [skillChainId];
exports.skills = skills;
