// saihubot-skill-diagnostics/dist/saihubot-skill-diagnostics.js
'use strict'; // plugin that provide ping, time, echo skills

/** simply return pong */

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.skills = exports.skillCurrentTime = exports.skillEcho = exports.skillPing = void 0;
const skillPing = {
  name: 'ping',
  help: 'ping - return pong',
  requirements: {},
  rule: /PING$/i,
  action: function (robot, msg) {
    robot.send('PONG');
  }
};
/** simply return anything you input */

exports.skillPing = skillPing;
const skillEcho = {
  name: 'echo',
  help: 'echo [string] - return [string]',
  requirements: {},
  rule: /ECHO (.*)$/i,
  action: function (robot, msg) {
    robot.send(msg[0]);
  }
};
/** return local time */

exports.skillEcho = skillEcho;
const skillCurrentTime = {
  name: 'time',
  help: 'time - return current browser time',
  requirements: {},
  rule: /TIME*|DATE*/i,
  action: function (robot, msg) {
    robot.send('Device time is ' + new Date().toLocaleString());
  }
};
exports.skillCurrentTime = skillCurrentTime;
const skills = [skillPing, skillEcho, skillCurrentTime];
exports.skills = skills;
